multi+refactor: persistent peer manager

ellemouton opened the pull request:
This PR adds a `PersistentPeerManager` to the server and refactors all persistent peer logic to be handled by the `PersistentPeerManager`. The end result is that  `persistentPeers`, `persistentPeersBackoff`, `persistentConnReqs` and `persistentRetryCancels` are all removed from the `server` struct and replaced by a `PersistentPeerManager`.

ellemouton commented on peer/connmgr.go:
```diff
@@ -0,0 +1,369 @@
+package peer
```
Note: I will still add unit tests for this. Just waiting for a go-ahead on the approach first

alexbosworth commented on lntest/itest/lnd_network_test.go:
```diff
@@ -60,6 +60,173 @@ func testNetworkConnectionTimeout(net *lntest.NetworkHarness, t *harnessTest) {
 }
 
+// testReconnectAfterIPChange verifies that if a persistent inbound node changes
+// its listening address then it's peer will still be able to reconnect to it.
```
```suggestion
// its listening address then its peer will still be able to reconnect to it.
```
```suggestion
// its listening address then it's peer will still be able to reconnect to it.
```

yyforyongyu commented on lntest/itest/lnd_network_test.go:
```diff
@@ -60,6 +60,149 @@ func testNetworkConnectionTimeout(net *lntest.NetworkHarness, t *harnessTest) {
+
+// withP2PPort is an optional function used to set the P2P port that a node
+// should use.
+func withP2PPort(port int) lntest.NodeOption {
```
how about making this into a helper closure?

yyforyongyu reviewed (commented):
cACK, I like the direction where we start moving code out of the `server.go`, def involves more work tho. My question is, since we are already here, can we move all connection management into peer?

ellemouton commented:
>  My question is, since we are already here, can we move all connection management into peer?

Very happy to do so! Do you think it should all be in 1 PR?

yyforyongyu commented:
> Mind elaborating? Do you mean in general or specifically in this PR?

Meant in this PR. Since we are moving the code outside of `server.go`, it's finally a good chance to patch the missing unit tests and refactor some of them if needed. But yeah, it would involve more work.

> Very happy to do so! Do you think it should all be in 1 PR?

It depends. I'd think about how emergent issue #5377 is. If it's a pressing issue, I'd suggest we move to #5538 and land it first, since it's almost done and has been reviewed. If not, we might as well take the chance to refactor the connection manager here, moving it into its own package. This will def take more time as the scale is large, but we should do it imo. Another option is to continue what you have here, that we move out the persistent manager as the first step. Guess it's a judgment call from @Roasbeef .

viaj3ro commented:
> If it's a pressing issue, I'd suggest we move to #5538 and land it first, since it's almost done and has been reviewed

I'd argue it is pressing. I still have a handful of TOR peers that refuse to connect back to me even after almost 4 months. Since many nodes have IP address changes once in a while, I assume they have the same issue without ever noticing it.

ellemouton commented:
holding off on this PR until 0.15. Working on the solution in #5538 again instead 👍

hsjoberg commented:
I'd also argue it's quite pressing, I have several users of my wallet which have encountered this issue.

lightninglabs-deploy commented:
@bhandras: review reminder
@roasbeef: review reminder
@yyforyongyu: review reminder

ellemouton commented:
!lightninglabs-deploy mute 2022-Jan-01

lightninglabs-deploy commented:
@bhandras: review reminder
@roasbeef: review reminder
@yyforyongyu: review reminder
@ellemouton, remember to re-request review from reviewers when ready

yyforyongyu commented:
Is this ready for another round?

ellemouton commented:
@yyforyongyu, yes it is ready :)

yyforyongyu reviewed (changes requested):
Good starting point! I'm thinking maybe we should add a dedicated package/service to act as our conn manager, but that's for the long run. I'm halfway through, submitting some questions first.

ellemouton commented on peer/persistentpeer_mgr.go:
```diff
@@ -0,0 +1,479 @@
+	conns map[string]*persistentPeer
+
+	// connsMu is used to guard the conns map.
+	connsMu sync.RWMutex
```
hmm but it does only guard the conns map at the moment..? happy to rename it to mu. but not sure about embedding it

ellemouton commented on peer/persistentpeer_mgr.go:
```diff
@@ -0,0 +1,479 @@
+// no open channels with them.
+type persistentPeer struct {
+	// addrs is all the addresses we know about for this peer.
+	addrs map[string]*lnwire.NetAddress
```
no it is the address string. will update comment to make it more clear 👍

ellemouton commented on peer/persistentpeer_mgr.go:
```diff
@@ -0,0 +1,479 @@
+
+	// retryCanceller is used to cancel any retry attempts with backoffs
+	// that are still maturing.
+	retryCanceller *chan struct{}
```
it is the only way i could get this initialisation logic to work

```
// Initialize a retry canceller for this peer if one does not exist.
	var cancelChan chan struct{}
	if peer.retryCanceller != nil {
		cancelChan = *peer.retryCanceller
	} else {
		cancelChan = make(chan struct{})
		peer.retryCanceller = &cancelChan
	}
```

Crypt-iQ replied:
I think you could use a regular channel and a boolean that, when set, means that the channel was closed. They both achieve the same thing, but I think the bool approach is nicer

ellemouton commented on peer/persistentpeer_mgr.go:
```diff
@@ -0,0 +1,479 @@
+	m.wg.Add(1)
+	go func() {
+		defer func() {
+			graphSub.Cancel()
```
because the subscription must not be canceled when `Start` exits, it must be canceled when the goroutine exits. Start will exit as soon as it has spun off the goroutine

ellemouton commented on peer/persistentpeer_mgr.go:
```diff
@@ -0,0 +1,479 @@
+				// If the router is shutting down, then we will
+				// as well.
+				if !ok {
+					return
```
I have added a log for now. 
not sure what the best way is to return errors from a goroutine? should the caller of `Start` be given an error channel to watch?

ellemouton commented on peer/persistentpeer_mgr.go:
```diff
@@ -0,0 +1,479 @@
+					return
+				}
+
+				for _, update := range topChange.NodeUpdates {
```
hehe good idea :)

ellemouton commented on peer/persistentpeer_mgr.go:
```diff
@@ -0,0 +1,479 @@
+						addrs[lnAddr.String()] = lnAddr
+					}
+
+					m.connsMu.Lock()
```
oh! thanks. updated 👍

ellemouton commented on peer/persistentpeer_mgr.go:
```diff
@@ -0,0 +1,479 @@
+		return
+	}
+
+	if len(peer.addrs) == 0 {
```
we might have added a persistent peer using `AddPeer` with a nil addr map. This is especially needed for the transition between the current state of the peer logic and this PR. But maybe yes, after the transition, we will expect addr to always be non-empty

ellemouton commented on peer/persistentpeer_mgr.go:
```diff
@@ -0,0 +1,479 @@
+	// addresses left over in the map will indicate which addresses we
+	// should create new connection requests for.
+	addrMap := make(map[string]*lnwire.NetAddress)
+	for _, addr := range peer.addrs {
```
not sure what you mean? this is a temporary map that we can freely delete things from in this function.

ellemouton commented on peer/persistentpeer_mgr.go:
```diff
@@ -0,0 +1,479 @@
+		// updatedConnReqs and remove the associated address from
+		// addrMap so that we don't recreate this connReq later on.
+		case true:
+			updatedConnReqs = append(updatedConnReqs, connReq)
```
I think we only ever want 1 connection. see https://github.com/lightning/bolts/blob/master/01-messaging.md#connection-handling-and-multiplexing

ellemouton commented on peer/persistentpeer_mgr.go:
```diff
@@ -0,0 +1,479 @@
+			// needed.
+			connReq := &connmgr.ConnReq{
+				Addr:      addr,
+				Permanent: true,
```
yes i think so. This is just a duplicate of the `connectToPersistentPeer` logic in `server.go`. That code was recently added but you can see in this diff [here](4d73f966aa68720542932437d688f56b370) that it was also set to `true` before

ellemouton commented on peer/persistentpeer_mgr.go:
```diff
@@ -0,0 +1,479 @@
+			}
+
+			m.connsMu.Lock()
+			if peer, ok = m.conns[pubKeyStr]; !ok {
```
yeah defs. but just added it since this is a goroutine so there is still a change we remove the peer from the map before getting here

ellemouton reviewed (commented):
Thanks for the initial review @yyforyongyu 🚀 
I have left some comments & addressed some of you questions and did some updates

lightninglabs-deploy commented:
@bhandras: review reminder
@roasbeef: review reminder
@yyforyongyu: review reminder
@ellemouton, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@bhandras: review reminder
@roasbeef: review reminder
@yyforyongyu: review reminder
@ellemouton, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@bhandras: review reminder
@roasbeef: review reminder
@yyforyongyu: review reminder
@ellemouton, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@bhandras: review reminder
@roasbeef: review reminder
@yyforyongyu: review reminder
@ellemouton, remember to re-request review from reviewers when ready

joostjager reviewed (commented):
Not totally familiar with this part of the code, but left a few comments. For me, the various goroutines and locks make it difficult to ascertain that the logic is correct. I wonder if it would be easier to understand with a single event loop per peer that receives updates via a channel.

lightninglabs-deploy commented:
@bhandras: review reminder
@roasbeef: review reminder
@yyforyongyu: review reminder
@ellemouton, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@bhandras: review reminder
@roasbeef: review reminder
@yyforyongyu: review reminder
@ellemouton, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@bhandras: review reminder
@roasbeef: review reminder
@yyforyongyu: review reminder
@ellemouton, remember to re-request review from reviewers when ready

ellemouton commented:
Hi @joostjager ! Thanks so much for the review on this! Im gonna mute the bot on this PR for a bit just while 0.15 is top prio & will address your review as soon as i can 👍

ellemouton commented:
!lightninglabs-deploy mute 2022-April-18

ellemouton commented on server.go:
```diff
@@ -2888,11 +2891,13 @@ func (s *server) prunePersistentPeerConnection(compressedPubKey [33]byte) {
-	if perm, ok := s.persistentPeers[pubKeyStr]; ok && !perm {
-		delete(s.persistentPeers, pubKeyStr)
+	if s.persistentPeerMgr.IsPersistentPeer(pubKeyStr) &&
+		!s.persistentPeerMgr.IsPermPeer(pubKeyStr) {
```
will add 👍

ellemouton commented on server.go:
```diff
@@ -1276,8 +1169,34 @@ func newServer(cfg *Config, listenAddrs []net.Addr,
+			if !s.persistentPeerMgr.IsPersistentPeer(pubStr) {
+				// Fetch any stored addresses we may have for
+				// this peer.
+				advertisedAddrs, err := s.fetchNodeAdvertisedAddrs(peerKey)
```
yeah good idea

ellemouton commented:
I will push an update addressing all the comments after the review club tomorrow (I decided to change the structure quite a bit so I don't want to push it before the review club in case people are already familiar with the old structure)

lightninglabs-deploy commented:
@bhandras: review reminder
@roasbeef: review reminder
@yyforyongyu: review reminder
@ellemouton, remember to re-request review from reviewers when ready

ellemouton commented on peer/persistentpeer_mgr.go:
```diff
@@ -0,0 +1,496 @@
+		ticker := time.NewTicker(m.cfg.MultiAddrConnectionStagger)
+		defer ticker.Stop()
+
+		for _, addr := range addrMap {
```
cool, added this in a later commit so that this one can remain a refactor

ellemouton commented on peer/persistentpeer_mgr.go:
```diff
@@ -0,0 +1,603 @@
+	// We choose not to wait group this go routine since the Connect call
+	// can stall for arbitrarily long if we shutdown while an outbound
+	// connection attempt is being made.
+	go func() {
```
I think this should be fine now that any previous retry canceller channel will get closed

lightninglabs-deploy commented:
@bhandras: review reminder
@roasbeef: review reminder
@yyforyongyu: review reminder
@ellemouton, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@bhandras: review reminder
@roasbeef: review reminder
@yyforyongyu: review reminder
@ellemouton, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@roasbeef: review reminder
@yyforyongyu: review reminder
@ellemouton, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@bhandras: review reminder
@roasbeef: review reminder
@yyforyongyu: review reminder
@crypt-iq: review reminder
@ellemouton, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@bhandras: review reminder
@roasbeef: review reminder
@yyforyongyu: review reminder
@crypt-iq: review reminder
@ellemouton, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@bhandras: review reminder
@roasbeef: review reminder
@yyforyongyu: review reminder
@crypt-iq: review reminder
@ellemouton, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@bhandras: review reminder
@roasbeef: review reminder
@yyforyongyu: review reminder
@crypt-iq: review reminder
@ellemouton, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@bhandras: review reminder
@roasbeef: review reminder
@yyforyongyu: review reminder
@crypt-iq: review reminder
@ellemouton, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@roasbeef: review reminder
@yyforyongyu: review reminder
@ellemouton, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@bhandras: review reminder
@roasbeef: review reminder
@yyforyongyu: review reminder
@crypt-iq: review reminder
@ellemouton, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@bhandras: review reminder
@roasbeef: review reminder
@yyforyongyu: review reminder
@crypt-iq: review reminder
@ellemouton, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@bhandras: review reminder
@roasbeef: review reminder
@yyforyongyu: review reminder
@crypt-iq: review reminder
@ellemouton, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@roasbeef: review reminder
@yyforyongyu: review reminder
@ellemouton, remember to re-request review from reviewers when ready

Crypt-iQ reviewed (commented):
checkpointing review here

lightninglabs-deploy commented:
@bhandras: review reminder
@roasbeef: review reminder
@yyforyongyu: review reminder
@ellemouton, remember to re-request review from reviewers when ready

joostjager commented on peer/persistentpeer_mgr.go:
```diff
@@ -0,0 +1,621 @@
+	// Next, check to see if we have any outstanding persistent connection
+	// requests to this peer. If so, then we'll remove all of these
+	// connection requests, and also delete the entry from the map.
+	if len(peer.connReqs) == 0 {
```
I think at this point, there may still be a `ConnectPeer` goroutine blocked on obtaining the lock. After this function returns and unblocks, `ConnectPeer` may still add something to the `connReqs`?

Crypt-iQ replied:
Yes this seems possible, if the goroutine spawned by `ConnectPeer` doesn't get canceled by the chan and is waiting on the mutex lock

joostjager commented on peer/persistentpeer_mgr.go:
```diff
@@ -0,0 +1,621 @@
+	FetchNodeAdvertisedAddrs func(route.Vertex) ([]net.Addr, error)
+}
+
+// PersistentPeerManager manages persistent peers.
```
Maybe add a basic description of what is/makes a peer persistent?

joostjager commented on peer/persistentpeer_mgr.go:
```diff
@@ -0,0 +1,621 @@
+	}
+
+	peer.connReqs = updatedConnReqs
+	cancelChan := peer.getRetryCanceller()
```
I think there may be a race condition with running goroutines created by the previous `ConnectPeer` that are blocked on the mutex?

Crypt-iQ replied:
with `peer.connReqs`?

joostjager commented on peer/persistentpeer_mgr.go:
```diff
@@ -0,0 +1,621 @@
+	// exist.
+	cancelChan := peer.getRetryCanceller()
+
+	// We choose not to wait group this go routine since the Connect call
```
Isn't the `ConnectPeer` call doing its connects in a goroutine?

joostjager reviewed (commented):
Did another pass of this PR focusing on concurrency issues first. I have to say that I find it hard to be certain that there are none with the extensive use of the lock, goroutines and cancel channel.

After my previous review, I mentioned:

>I wonder if it would be easier to understand with a single event loop per peer that receives updates via a channel.

I still think that would help a lot. Not just for review, but also for future devs working on this code.

Generally speaking, the connection handling for a peer is independent of all other peers. If that would be reflected in the code as a layer, it's already a bit easier to see how things work.

Then within the code for handling a single peer, I think that a single event loop that deals with retries, new addresses, backoff and additional connection requests will make the logic much more transparent.

lightninglabs-deploy commented:
@bhandras: review reminder
@roasbeef: review reminder
@yyforyongyu: review reminder
@ellemouton, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@bhandras: review reminder
@roasbeef: review reminder
@yyforyongyu: review reminder
@ellemouton, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@bhandras: review reminder
@roasbeef: review reminder
@yyforyongyu: review reminder
@ellemouton, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@bhandras: review reminder
@roasbeef: review reminder
@yyforyongyu: review reminder
@ellemouton, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@bhandras: review reminder
@roasbeef: review reminder
@yyforyongyu: review reminder
@ellemouton, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@bhandras: review reminder
@roasbeef: review reminder
@yyforyongyu: review reminder
@ellemouton, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@bhandras: review reminder
@roasbeef: review reminder
@yyforyongyu: review reminder
@ellemouton, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@bhandras: review reminder
@roasbeef: review reminder
@yyforyongyu: review reminder
@ellemouton, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@bhandras: review reminder
@roasbeef: review reminder
@yyforyongyu: review reminder
@ellemouton, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@bhandras: review reminder
@roasbeef: review reminder
@yyforyongyu: review reminder
@ellemouton, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@bhandras: review reminder
@roasbeef: review reminder
@yyforyongyu: review reminder
@ellemouton, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@bhandras: review reminder
@roasbeef: review reminder
@yyforyongyu: review reminder
@ellemouton, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@bhandras: review reminder
@roasbeef: review reminder
@yyforyongyu: review reminder
@ellemouton, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@bhandras: review reminder
@roasbeef: review reminder
@yyforyongyu: review reminder
@ellemouton, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@bhandras: review reminder
@roasbeef: review reminder
@yyforyongyu: review reminder
@ellemouton, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@bhandras: review reminder
@roasbeef: review reminder
@yyforyongyu: review reminder
@ellemouton, remember to re-request review from reviewers when ready

ellemouton commented:
!lightninglabs-deploy mute

ellemouton commented:
muting for a bit. Will pick this up again when my plate clears up a bit

ellemouton commented:
closing for now. Can re-open once re-prio'd
